File: nugurang-back/src/main/java/com/nugurang/graphql/ThreadResolver.java
package com.nugurang.graphql;

import com.nugurang.dao.ArticleDao;
import com.nugurang.dao.BoardDao;
import com.nugurang.dao.EventDao;
import com.nugurang.dao.TeamDao;
import com.nugurang.dao.ThreadDao;
import com.nugurang.dao.UserDao;
import com.nugurang.dto.ArticleDto;
import com.nugurang.dto.BoardDto;
import com.nugurang.dto.EventDto;
import com.nugurang.dto.TeamDto;
import com.nugurang.dto.ThreadDto;
import com.nugurang.dto.UserDto;
import graphql.kickstart.tools.GraphQLResolver;
import java.util.List;
import java.util.stream.Collectors;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.PageRequest;
import org.springframework.stereotype.Service;

@RequiredArgsConstructor
@Service
public class ThreadResolver implements GraphQLResolver<ThreadDto> {
    private final ArticleDao articleDao;
    private final BoardDao boardDao;
    private final TeamDao teamDao;
    private final ThreadDao threadDao;
    private final UserDao userDao;
    private final EventDao eventDao;
    private final ArticleDao articledDao;

    public BoardDto board(ThreadDto threadDto) {
        return threadDao
            .findById(threadDto.getId())
            .map((threadEntity) ->
                threadEntity.getBoard()
            )
            .map((boardEntity) ->
                BoardDto.builder()
                .id(boardEntity.getId())
                .name(boardEntity.getName())
                .build()
            ).get();
    }

    public UserDto user(ThreadDto threadDto) {
        return null;
    }

    public TeamDto team(ThreadDto threadDto) {
        return null;
    }

    public EventDto event(ThreadDto threadDto) {
        return null;
    }

    public List<ArticleDto> getArticles(ThreadDto threadDto, Integer page, Integer pageSize) {
        return articleDao
            .findAllByThreadIdOrderByCreatedAtAsc(threadDto.getId(), PageRequest.of(page, pageSize))
            .getContent()
            .stream()
            .map((entity) -> entity.toDto())
            .collect(Collectors.toList());
    }

}

File: src/main/kotlin/com/nugurang/graphql/resolver/ThreadResolver.kt
package com.nugurang.graphql.resolver

import com.nugurang.dao.ArticleDao
import com.nugurang.dao.ThreadDao
import com.nugurang.dto.*
import com.nugurang.entity.BoardEntity
import com.nugurang.entity.UserEntity
import com.nugurang.exception.NotFoundException
import com.nugurang.mapper.ArticleMapper
import com.nugurang.mapper.BoardMapper
import com.nugurang.mapper.EventMapper
import com.nugurang.mapper.UserMapper
import com.nugurang.service.ThreadService
import graphql.kickstart.tools.GraphQLResolver
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service

@Service
class ThreadResolver(
    private val articleDao: ArticleDao,
    private val threadDao: ThreadDao,
    private val threadService: ThreadService,
    private val articleMapper: ArticleMapper,
    private val boardMapper: BoardMapper,
    private val eventMapper: EventMapper,
    private val userMapper: UserMapper
) : GraphQLResolver<ThreadDto> {
    fun board(threadDto: ThreadDto): BoardDto {
        val board = threadDao.findByIdOrNull(threadDto.id)?.board ?: throw NotFoundException(BoardEntity::class.java)
        return boardMapper.toDto(board)
    }

    fun user(threadDto: ThreadDto): UserDto {
        return threadDao
        .findByIdOrNull(threadDto.id)
        ?.user
        ?.let(userMapper::toDto)
        ?: throw NotFoundException(UserEntity::class.java)
    }

    fun team(threadDto: ThreadDto): TeamDto? {
        return null
    }

    fun event(threadDto: ThreadDto): EventDto? {
        return threadDao.findByIdOrNull(threadDto.id)?.event?.let(eventMapper::toDto)
    }

    fun tags(threadDto: ThreadDto): List<TagDto> {
        return listOf()
    }

    fun firstArticle(threadDto: ThreadDto): ArticleDto {
        return articleMapper.toDto(threadService.getFirstArticle(threadDto.id))
    }

    fun getArticles(threadDto: ThreadDto, page: Int, pageSize: Int): List<ArticleDto> {
        return threadService.getArticles(threadDto.id, page, pageSize).map(articleMapper::toDto)
    }

    fun commentCount(threadDto: ThreadDto): Long {
        return articleDao.countByThreadId(threadDto.id) - 1
    }

    fun test(t: Test?) {
        val x = t?.x
        println(x!!)
    }
}
class Test(val x: Int)